File: backend/src/modules/carrierroute/carrierroute.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { CarrierName } from './entities/carrier-name.entity';
import { DomainName } from './entities/domain-name.entity';
import { CarrierFailureRoute } from './entities/carrierfailureroute.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';
import { KamailioRpcService } from '../../common/services/kamailio-rpc.service';

@Injectable()
export class CarrierrouteService {
  constructor(
    @InjectRepository(CarrierName)
    private carrierNameRepository: Repository<CarrierName>,
    @InjectRepository(DomainName)
    private domainNameRepository: Repository<DomainName>,
    @InjectRepository(CarrierFailureRoute)
    private failureRouteRepository: Repository<CarrierFailureRoute>,
    private kamailioRpcService: KamailioRpcService,
  ) {}

  // Carrier Names
  async findAllCarriers(): Promise<CarrierName[]> {
    return this.carrierNameRepository.find({ order: { id: 'ASC' } });
  }

  async createCarrier(carrier: string): Promise<CarrierName> {
    return this.carrierNameRepository.save(this.carrierNameRepository.create({ carrier }));
  }

  async deleteCarrier(id: number): Promise<void> {
    await this.carrierNameRepository.delete(id);
  }

  // Domain Names
  async findAllDomains(): Promise<DomainName[]> {
    return this.domainNameRepository.find({ order: { id: 'ASC' } });
  }

  async createDomain(domain: string): Promise<DomainName> {
    return this.domainNameRepository.save(this.domainNameRepository.create({ domain }));
  }

  async deleteDomain(id: number): Promise<void> {
    await this.domainNameRepository.delete(id);
  }

  // Failure Routes
  async findAllFailureRoutes(dto: PaginationDto): Promise<PaginatedResult<CarrierFailureRoute>> {
    const { page, limit } = dto;
    const skip = (page - 1) * limit;
    const [items, total] = await this.failureRouteRepository.findAndCount({
      skip, take: limit, order: { id: 'ASC' },
    });
    return new PaginatedResult(items, total, page, limit);
  }

  async createFailureRoute(data: Partial<CarrierFailureRoute>): Promise<CarrierFailureRoute> {
    return this.failureRouteRepository.save(this.failureRouteRepository.create(data));
  }

  async deleteFailureRoute(id: number): Promise<void> {
    await this.failureRouteRepository.delete(id);
  }

  // RPC
  async reload(): Promise<void> {
    await this.kamailioRpcService.call('cr.reload_routes');
  }

  async dumpRoutes(): Promise<any> {
    return this.kamailioRpcService.call('cr.dump_routes');
  }
}

File: kam-admin-module-routing/src/main/java/io/kamailio/admin/modules/carrierroute/service/CarrierrouteService.java
package io.kamailio.admin.modules.carrierroute.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.common.service.KamailioRpcService;
import io.kamailio.admin.modules.carrierroute.entity.CarrierFailureRoute;
import io.kamailio.admin.modules.carrierroute.entity.CarrierName;
import io.kamailio.admin.modules.carrierroute.entity.DomainName;
import io.kamailio.admin.modules.carrierroute.repository.CarrierFailureRouteRepository;
import io.kamailio.admin.modules.carrierroute.repository.CarrierNameRepository;
import io.kamailio.admin.modules.carrierroute.repository.DomainNameRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class CarrierrouteService {
    private final CarrierNameRepository carrierRepository;
    private final DomainNameRepository domainRepository;
    private final CarrierFailureRouteRepository failureRouteRepository;
    private final KamailioRpcService kamailioRpc;

    public CarrierrouteService(CarrierNameRepository carrierRepository, DomainNameRepository domainRepository,
                               CarrierFailureRouteRepository failureRouteRepository, KamailioRpcService kamailioRpc) {
        this.carrierRepository = carrierRepository;
        this.domainRepository = domainRepository;
        this.failureRouteRepository = failureRouteRepository;
        this.kamailioRpc = kamailioRpc;
    }

    public List<CarrierName> findAllCarriers() {
        return carrierRepository.findAll(Sort.by("id"));
    }

    @Transactional
    public CarrierName createCarrier(String carrier) {
        var e = new CarrierName();
        e.setCarrier(carrier);
        return carrierRepository.save(e);
    }

    @Transactional
    public void deleteCarrier(Integer id) { carrierRepository.deleteById(id); }

    public List<DomainName> findAllDomains() {
        return domainRepository.findAll(Sort.by("id"));
    }

    @Transactional
    public DomainName createDomain(String domain) {
        var e = new DomainName();
        e.setDomain(domain);
        return domainRepository.save(e);
    }

    @Transactional
    public void deleteDomain(Integer id) { domainRepository.deleteById(id); }

    public PaginatedResult<CarrierFailureRoute> findAllFailureRoutes(PaginationDto dto) {
        var page = failureRouteRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public CarrierFailureRoute createFailureRoute(CarrierFailureRoute data) { return failureRouteRepository.save(data); }

    @Transactional
    public void deleteFailureRoute(Integer id) { failureRouteRepository.deleteById(id); }

    public void reload() { kamailioRpc.crReloadRoutes().block(); }
    public Object dumpRoutes() { return kamailioRpc.crDumpRoutes().block(); }
}
